middleware/common/source/service/lookup.cpp

```diff
@@ -12,145 +12,111 @@ namespace casual
-               lookup::discard( m_correlation);
-         }
+                  if( reply.state == decltype( reply.state)::replied)
+                     communication::ipc::inbound::device().discard( correlation);
+               });
+            }
 
-         Lookup::Lookup( Lookup&& other) noexcept
-            : m_service{ std::exchange( other.m_service, {})},
-               m_correlation{ std::exchange( other.m_correlation, {})},
-               m_reply{ std::exchange( other.m_reply, {})}
-         {}
+            void validate( const lookup::Reply& reply, const std::string& service)
+            {
+               using Enum = decltype( reply.state);
+               switch( reply.state)
+               {
+                  case Enum::idle:
+                     return;
+                  case Enum::absent:
+                     code::raise::error( code::xatmi::no_entry, "failed to lookup service: ", service);
+                  case Enum::timeout:
+                     code::raise::error( code::xatmi::timeout, "timeout during lookup of service: ", service);
+               };
+            }
+            
+         } // <unnamed>
+      } // local
 
-         const strong::correlation::id& Lookup::correlation() const noexcept
-         {
-            if( m_reply)
-               return m_reply->correlation;
+      Lookup::Lookup( std::string service, lookup::Context context, std::optional< platform::time::point::type> deadline)
+         : m_service( std::move( service))
+      {
+         Trace trace{ "common::service::Lookup"};
 
-            return m_correlation;
-         }
+         message::service::lookup::Request request{ process::handle()};
+         request.requested = m_service;
+         request.context = context;
+         request.deadline = std::move( deadline);
 
-         Lookup& Lookup::operator = ( Lookup&& other) noexcept
-         {
-            m_service = std::exchange( other.m_service, {});
-            m_correlation = std::exchange( other.m_correlation, {});
-            m_reply = std::exchange( other.m_reply, {});
-            return *this;
-         }
+         m_correlation = communication::device::blocking::send( communication::instance::outbound::service::manager::device(), request);
+      }
 
-         void swap( Lookup& lhs, Lookup& rhs)
-         {
-            using std::swap;
-            swap( lhs.m_service, rhs.m_service);
-            swap( lhs.m_reply, rhs.m_reply);
-            swap( lhs.m_correlation, rhs.m_correlation);
-         }
+      Lookup::Lookup( std::string service, std::optional< platform::time::point::type> deadline) 
+         : Lookup( std::move( service), lookup::Context{}, std::move( deadline)) 
+      {}
 
-         void Lookup::update( Reply&& reply)
-         {
-            log::line( verbose::log, "reply: ", reply);
+      Lookup::~Lookup()
+      {
+         // we have to discard if we're still pending with the service manager
+         if( ! m_correlation)
```

Comment:
If m_correlation is empty I dont think it makes sense to call local::discard since that function has a guard that returns if the correlation is empty? Should it be `if( m_correlation)`?

Reply:
Of course!